### PR TITLE
Issue 106/investigate canonical telemetry first event model

### LINKED ISSUE #106: [Feature] Investigate canonical telemetry-first event model across runtime, sink, and TUI
## Summary
Investigate moving Stormlog toward one primary event model across runtime capture, append-only sink persistence, artifact loading, and live/TUI display.

## Why
Stormlog currently has multiple event shapes and normalization paths. That works, but it makes future queryability, aggregation, integrations, and long-term schema evolution more expensive than they need to be.

## Research Questions
- Should runtime trackers emit canonical telemetry records directly, or keep a lighter internal event and normalize at the edge?
- What hot-path overhead is acceptable for live tracking and always-on mode?
- How should CPU, GPU, and future TPU/backends share one contract without overfitting to one runtime?
- What compatibility layer is needed for legacy exports, tests, and older artifact formats?
- How should live sessions and artifact-loaded sessions expose the same data model to the TUI and analysis code?

## Proposed Scope
- Compare the current split between runtime tracking events and canonical telemetry records.
- Define the smallest event contract that can power runtime, sink, TUI, and exported artifacts.
- Identify migration points in trackers, sink writers, loaders, and TUI adapters.
- Document compatibility and performance tradeoffs.

## Success Criteria
- A written design for a telemetry-first internal model.
- A migration plan that preserves backward compatibility for existing artifact formats.
- A clear benchmark plan for runtime overhead and allocation churn.
- Follow-on implementation tasks split by tracker, sink, loader, and UI layer.

## Related Links
- #32
- #33
- OpenTelemetry Signals: https://opentelemetry.io/docs/concepts/signals/
- OpenTelemetry Trace Overview: https://opentelemetry.io/docs/reference/specification/overview/

### PR DESCRIPTION
**Pull request by nanaagyei**
## Summary

Implements the telemetry-first internal model work by adding a canonical, backend-neutral telemetry projection over the existing `TelemetryEvent v3` contract and documenting the architecture decision in the project docs.

This keeps the current artifact and sink formats backward-compatible while giving runtime, loaded artifacts, live/TUI sessions, and future analysis code a shared internal event model.

## Changes

- Added `CanonicalTelemetryRecord` and projection helpers in `stormlog.telemetry_model`.
- Added canonical record helpers in `stormlog.telemetry`.
- Exposed canonical telemetry records from:
  - `LoadedTelemetrySession.telemetry_records()`
  - `LoadedTelemetrySession.resources()`
  - `LoadedTelemetrySession.correlations()`
  - `TrackerSession.telemetry_records()`
- Preserved existing V2/V3 artifact loading and current sink/export behavior.
- Added tests for canonical projection, deterministic record IDs, immutability, legacy mapping normalization, loaded sessions, and TUI session projection.
- Added `docs/telemetry_first_internal_model.md`.
- Linked the new guide from the docs index and API/reference docs.
- Cross-referenced the current telemetry schema without changing the schema itself.
- Cleaned Sphinx API navigation duplication.

## Design Notes

- Trackers can continue emitting compact runtime-local events.
- Normalization remains at the tracker/loader edge.
- `TelemetryEvent v3` remains the persisted artifact and append-only sink format for now.
- The canonical model is an additive internal projection, not a breaking schema migration.
- Backend-specific details live in `resource`, `attributes`, and `correlation` rather than top-level backend-specific fields.
- Future persisted canonical writes are documented as a separate migration phase gated by benchmark results.

## Compatibility

- Existing V2, V3, legacy JSON, and JSONL artifacts remain loadable.
- No runtime public API behavior was removed.
- No existing JSON schema was changed.
- Legacy export and loader paths remain explicit compatibility boundaries.

## Validation

Passed locally with `.venv`:

- `pre-commit run --all-files`
- `python -m black --check stormlog/ tests/ examples/`
- `python -m isort --check --diff stormlog/ tests/ examples/`
- `python -m flake8 stormlog/ tests/ examples/ --show-source --statistics`
- `python -m mypy stormlog/`
- `python -m sphinx -W --keep-going -b html docs docs/_build/html`
- `python -m pytest tests/ -m "not tui_pilot and not tui_snapshot and not tui_pty" -q`
- `python -m pytest tests/tui/ -m "tui_pilot or tui_snapshot" -q`
- `python -m pytest tests/e2e/test_tui_pty.py -m tui_pty -q`
- `python -m build --outdir /private/tmp/stormlog-dist`
- `python -m twine check /private/tmp/stormlog-dist/*`

Benchmark gate note:

- CPU gpumemprof regression metrics passed.
- Full benchmark command failed only because TensorFlow is not installed in the local `.venv`.

Closes #106

### CODE REVIEW

File: docs/telemetry_first_internal_model.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
+[← Back to main docs](index.md)
+
+# Stormlog Telemetry-First Internal Model
```
Comment (Silas-Asamoah) [inline docs/telemetry_first_internal_model.md]: Issue: the title and filename read like a phase-specific design memo rather than durable documentation. `telemetry_first_internal_model` also leaves room to imply a later second internal model. We could rename this to something stable such as `telemetry_internal_model`, `telemetry_projection`, or place it under an architecture/ADR-style section if the goal is to preserve the rationale.

File: docs/telemetry_schema.md [hunk 2]
```diff
@@ -4,6 +4,9 @@
 
 `TelemetryEvent v3` is the canonical event format for tracker exports,
 append-only sink segments, synthesized diagnose timelines, and loader output.
+For the backend-neutral internal projection over this stable artifact format,
```
Comment (Silas-Asamoah) [inline docs/telemetry_schema.md]: Issue: this page already describes `TelemetryEvent v3` as the canonical event format, while the new model introduces `CanonicalTelemetryRecord`. That gives us two canonical concepts: the persisted artifact schema and the internal projection. We should either avoid `canonical` for the internal projection or explicitly distinguish persisted canonical schema from internal projected envelope throughout the docs.

File: stormlog/telemetry.py [hunk 3]
```diff
@@ -1142,14 +1163,41 @@ def load_telemetry_events(
...
 __all__ = [
     "SCHEMA_VERSION_V2",
     "SCHEMA_VERSION_V3",
     "SCHEMA_VERSION_LATEST",
+    "CanonicalTelemetryRecord",
     "LoadedTelemetrySession",
     "TelemetryEvent",
     "TelemetryEventV2",
     "TelemetryEventV3",
+    "canonical_record_from_telemetry_event",
```
Comment (Silas-Asamoah) [inline stormlog/telemetry.py]: Issue: exporting `canonical_*` helpers from `stormlog.telemetry` is convenient, but the prefix makes this look like the primary telemetry contract rather than a projection over `TelemetryEvent v3`. We could use projection-oriented names such as `project_telemetry_event` / `project_telemetry_events`, or document that these exports are an internal-model convenience while V3 remains the artifact contract.

File: stormlog/telemetry_model.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    "collector_recovered": "info",
+    "start": "info",
+    "stop": "info",
+    "phase_enter": "info",
+    "phase_exit": "info",
+    "sample": "info",
+}
+
+
+@dataclass(frozen=True)
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py:35]: Issue: `CanonicalTelemetryRecord` overlaps with the existing docs that call `TelemetryEvent v3` canonical. This can confuse which shape is stable externally versus internal-only. A name like `TelemetryEnvelope`, `InternalTelemetryRecord`, or `ProjectedTelemetryRecord` would better communicate that this is a backend-neutral projection over the persisted event schema.

File: stormlog/telemetry_model.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+from __future__ import annotations
+
+import hashlib
+import json
+from collections.abc import Mapping as MappingABC
+from dataclasses import dataclass, field
+from types import MappingProxyType
+from typing import Any, Iterable, Literal, Mapping, Optional, cast
+
+CANONICAL_TELEMETRY_SCHEMA_VERSION: Literal[1] = 1
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py]: Issue: using `Literal[1]` is fine for the current projection version, but the PR should define the bump procedure. A future version change needs to update this constant, the dataclass annotation, serialization tests, docs, and compatibility behavior together. Without that policy, the type narrowness can become stale when the projection schema evolves.

File: stormlog/telemetry_model.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    attributes: Mapping[str, Any] = field(default_factory=dict)
+    correlation: Mapping[str, Any] = field(default_factory=dict)
+
+
+def _freeze_value(value: Any) -> Any:
+    if isinstance(value, MappingABC):
+        return MappingProxyType(
+            {str(key): _freeze_value(nested) for key, nested in value.items()}
+        )
+    if isinstance(value, list):
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py]: Issue: `_freeze_value` recursively freezes lists but leaves input tuples untouched, while `_thaw_value` later treats tuples as lists. If metadata contains a tuple with a nested dict, the frozen record can still contain mutable nested state. We should either recursively freeze tuples here too, or state and validate that projection input must already be JSON-shaped.

File: stormlog/telemetry_model.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+def _jsonable_mapping(payload: Mapping[str, Any]) -> dict[str, Any]:
+    result = json.loads(json.dumps(_thaw_value(payload), sort_keys=True, default=str))
+    return cast(dict[str, Any], result)
+
+
+def _record_id(record: Mapping[str, Any]) -> str:
+    digest = hashlib.sha256(
+        json.dumps(_jsonable_mapping(record), sort_keys=True).encode("utf-8")
+    ).hexdigest()
+    return f"telemetry-{digest[:32]}"
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py:90]: Issue: truncating the SHA-256 digest to 32 hex characters is a reasonable compact 128-bit identifier, but the choice is currently implicit. Please add a small comment or test that records why 32 characters is sufficient so this does not look arbitrary when record identity becomes part of the internal contract.

File: stormlog/telemetry_model.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+def _source_kind(record: Mapping[str, Any], metadata: Mapping[str, Any]) -> str:
+    backend = metadata.get("backend")
+    if isinstance(backend, str) and backend.strip():
+        return backend.strip().lower()
+
+    collector = record.get("collector")
+    if isinstance(collector, str):
+        lowered = collector.lower()
+        for candidate in ("cuda", "rocm", "mps", "cpu", "tensorflow", "tpu"):
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py]: Issue: `tpu` is included as a source kind even though Stormlog does not currently support TPU capture. That makes the backend contract look broader than the implementation. We should either remove it until there is collector support, or explicitly document it as a reserved future value and keep tests from implying current TPU support.

File: stormlog/telemetry_model.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    explicit = metadata.get("severity")
+    if isinstance(explicit, str) and explicit.strip():
+        return explicit.strip().lower()
+    event_type = record.get("event_type")
+    if isinstance(event_type, str):
+        return _SEVERITY_BY_EVENT_TYPE.get(event_type.strip().lower())
+    return None
+
+
+def _resource(record: Mapping[str, Any], source_kind: str) -> dict[str, Any]:
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py:138]: Issue: `source_kind`, session/job/rank fields, and device identity are projected into multiple places across top-level fields, `resource`, and `correlation`. Duplication can be useful for indexing and grouping, but the contract should say these values are derived from the same source and must not drift. Otherwise future adapters may update one copy and leave another inconsistent.

File: stormlog/telemetry_model.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+@dataclass(frozen=True)
+class CanonicalTelemetryRecord:
+    """Small immutable event envelope shared by live and loaded telemetry."""
+
+    schema_version: Literal[1]
+    record_id: str
+    timestamp_ns: int
+    observed_timestamp_ns: int
+    session_id: str
+    source_kind: str
```
Comment (Silas-Asamoah) [inline stormlog/telemetry_model.py:44]: Issue: `source_kind` should stay a backend family, not a node or distributed identity. For multi-node GPU runs this should remain `cuda`/`rocm`/`mps`/`cpu`, while host, device id, rank, local rank, world size, and any future node id live in `resource` or `correlation`. Please document this so distributed adapters do not overload `source_kind`.

File: docs/telemetry_first_internal_model.md [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+2. Existing normalizers produce `TelemetryEvent v3` records for exports, sink
+   writes, loaders, and TUI adapters.
+3. `canonical_record_from_telemetry_event(...)` projects V3 records into
+   `CanonicalTelemetryRecord`.
+4. `LoadedTelemetrySession.telemetry_records()` exposes canonical records for
+   loaded artifacts.
+5. `TrackerSession.telemetry_records()` exposes canonical records for live TUI
+   sessions.
+
+This keeps the capture path cheap while giving downstream code one stable
```
Comment (Silas-Asamoah) [inline docs/telemetry_first_internal_model.md]: Issue: this statement is accurate for the current additive implementation because projection happens after existing V3 normalization, but it should be framed that way. Future sink migration that persists the envelope directly is not proven cheap until benchmark gates cover capture latency, allocations, queue depth, and sink throughput.

File: docs/telemetry_projection.md [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+Capture code should avoid:
+
+- blocking on sink persistence,
+- heap-heavy serialization,
+- reflection-heavy shaping,
+- formatting strings before normalization when raw fields are enough.
+
+Sink persistence should remain batched and append-only. Under pressure,
+Stormlog should prefer explicit sampling or low-priority event shedding over
```
Comment (Silas-Asamoah) [inline docs/telemetry_projection.md:176]: Issue: `under pressure` and `low-priority event shedding` need a concrete policy. Pressure could mean bounded history overflow, sink queue backlog, slow disk, bursty alert events, or long always-on sessions. If Stormlog sheds instead of blocking, the contract should require visible counters and reasons such as dropped samples, dropped events, queue depth, and the active sampling/shedding policy.

**Review by Silas-Asamoah — COMMENTED** [review]
The additive projection direction makes sense: keeping TelemetryEvent v3 as the persisted artifact while projecting into an internal backend-neutral envelope is a reasonable migration step. The main things to tighten are the naming/contract boundaries, source-kind scope, duplicated identity fields, pressure policy, and schema/versioning procedure. Specific suggestions are inline.

**Review by Silas-Asamoah — APPROVED** [review]
